### ISSUE TITLE
Implement Table Range locks for FaTE transactions

### ISSUE THREAD

**Issue by dlmarion** [apache/accumulo#2483]
**Is your feature request related to a problem?
The current table locks do not allow FaTE transactions operating on different parts of a table to run concurrently.

**Describe the solution you'd like**
Allow for different FaTE transactions to be able to operate on a table concurrently when they affect different ranges

User feedback for this issue captured at https://github.com/apache/accumulo/pull/2467#discussion_r801868052

**Comment by dlmarion** [comment]
@keith-turner @cshannon - Is this still an issue after the recent Fate changes? Can multiple Fate transactions run concurrently on different parts of the table? I'm thinking the answer is yes with the new operation_id column in the tablet metadata in v4.0.

**Comment by cshannon** [comment]
Yeah I would say that the opid column is the table range lock as it allows locking tablets for fate operations. Some fate operations stll require locking the entire tablet (like delete table, clone, etc) but there is no getting around that. I can let @keith-turner comment too but I believe operations like Split, merge, etc that set that operation id should be able to work concurrently.

**Comment by keith-turner** [comment]
I looked into removing the zookeeper table locks and ran into two problems.  

 1. Multi tablet operations like merge set operation ids on tablet independently.  Removing table locks for these would require more complex locking operations on the tablets, something resembling two phase commit.
 2. Some table data is stored in zookeeper, like table state.  So coordinating this type of zookeeper state w/ tablet state is much easier with the current zookeeper table locks.

The split code in accumulo 4 avoids using table locks, but that required adding complex code to coordinate with the table state stored in zookeeper.

The biggest advantage for table locks having a range would be allowing merge to run on one part of table and bulk import and compaction to continue to run on another part of the table.  But now that merge operations are much faster  this use case is not as compelling.  Compactions and bulk imports get read locks on tables, so those can run concurrently.

Looking through all of the table operations that get write locks, found the following.

 * create table
 * delete table
 * rename table
 * change table state (offline/online)
 * clone table
 * import table
 * merge

Merge seems to be the only operation that would benefit from a range on the table lock.  Create,delete, clone, etc would all to lock the full range of the table if locks had a range.  Since merge is faster now and does not need the ranges, maybe there is no good use case for this feature.

There is a caveat though.  Even though merge is faster, once a merge is initiated it will wait for running bulk imports and table compactions to complete.  While its waiting it will prevent new ones from starting.  So it could cause a disturbance in throughput for.

**Comment by keith-turner** [comment]
Adding ranges to the table locks would be useful.  This will help avoid a situation like the following.

 1. Long running table compaction over range (abc,def] is started and gets a read table lock
 2. Merge is started on range (prq,xyz] which queues up a write lock entry and waits for the existing read lock
 3. Bulk import is initiated on range (fg,mn], tries to get a read lock and blocks because a write lock is queued.

If the locks had ranges, then the bulk import could proceed because its range does not overlap w/ the queued write lock.

**Comment by keith-turner** [comment]
Implementing this may be fairly straightforward.  Currently locks do the following

 1. Read locks check for any write locks that sort earlier to determine if they hold the lock
 2. Write locks check for and read or write locks that sort earlier to determine if they hold the lock

Seems like the lock algorithm could change to the following.

 1. When a lock is added to zookeeper a range is also provided.
 2. Read locks check for any write locks with an overlapping range that sort earlier to determine if they hold the lock
 3. Write locks check for and read or write locks with an overlapping range that sort earlier to determine if they hold the lock

The above changes could be made initially and then follow on changes could be made to fate operations to use those changes.

**Comment by keith-turner** [comment]
Related issue #1324

**Comment by keith-turner** [comment]
Looked into adding the range from a fate operation like compact to the table lock and ran into one issue.  Some operations (like compact) find tablets that overlap the range specified by the API call.   For example a compaction initiated on range `(c,m]` would overlap tablet `(g,n]`.    If the compact operation associated the ragne `(c,m]` with its table lock, then it would not cover the full row range of data the operation is working on, so this will need to be considered.  Probably just need to always use existing tablet splits in the table lock ranges.

Could do the following for fate operations that have a range and currently work on tablets that overlap.

 1. Expand the row range supplied to the API to current tablets that it overlaps
 2. Lock using the expanded range
 3. Modify the fate operation to only operate on tablets that fall entirely within the range it locked.  This will make the operation resilient to race condition where the table splits change between step 1 and 2.  Would need to go back to step 1 if the set of tablets that fall completely within the locked range is an empty set.

**Comment by keith-turner** [comment]
Looking into this.  W/o this feature the new automatic merge feature will not work very well.
